### PR TITLE
STYLE: moving spacing conventions in cython casting to pre-commit

### LINKED ISSUE #42121: STYLE move spacing conventions in cython casting to pre-commit
here's the check which needs moving, it's in `ci/code_checks.sh`
```
### LINTING ###
if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then

    # Check that cython casting is of the form `<type>obj` as opposed to `<type> obj`;
    # it doesn't make a difference, but we want to be internally consistent.
    # Note: this grep pattern is (intended to be) equivalent to the python
    # regex r'(?<![ ->])> '
    MSG='Linting .pyx code for spacing conventions in casting' ; echo $MSG
    invgrep -r -E --include '*.pyx' --include '*.pxi.in' '[a-zA-Z0-9*]> ' pandas/_libs
    RET=$(($RET + $?)) ; echo $MSG "DONE"

    # readability/casting: Warnings about C casting instead of C++ casting
    # runtime/int: Warnings about using C number types instead of C++ ones
    # build/include_subdir: Warnings about prefacing included header files with directory

fi
```

It could be a `pygrep` check

### PR DESCRIPTION
**Pull request by 01-vyom**
- [X] closes #42121 
- [ ] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -102,6 +102,9 @@ repos:
             # Incorrect code-block / IPython directives
             |\.\.\ code-block\ ::
             |\.\.\ ipython\ ::
+
+            # Cython casting is of the form `<type>obj` as opposed to `<type> obj`
+            |[a-zA-Z0-9*]>
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:107]: is this correct? can you try writing something like
```
<type> obj
```
in a `.pyx` file to see if this check picks it up?

---

In the future, feel free to tag the author of an issue for review - there's so many PRs it's easy to miss them

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 1 above)
Comment (01-vyom) [reply .pre-commit-config.yaml:107]: Sorry I completely forgot about this, I was facing a problem in inserting the regex in the .yaml file.
The original regex has a 'space' at the end: `'[a-zA-Z0-9*]> '`, but when I tried to insert the same regex in the .yaml file, the pre-commit for black formatted it and removed the space at the end. Thus, it is not working properly. I also tested it on a `test.pyx` file with the following content: `<type> obj` as you suggested and the hook failed. 
Fail Message: 
```
> pre-commit run unwanted-patterns --files pandas/_libs/test.pyx
Unwanted patterns........................................................Failed
- hook id: unwanted-patterns
- exit code: 1

pandas/_libs/test.pyx:1:<test> obj
```

**Comment by 01-vyom** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.

Working

File: .pre-commit-config.yaml (hunk 1 above)
Comment (MarcoGorelli) [reply .pre-commit-config.yaml:107]: you might need to escape the whitespace, like `\ `

**Comment by MarcoGorelli** [comment]
are you sure? looks like there's still lot's of failures https://github.com/pandas-dev/pandas/pull/42186/checks?check_run_id=3164816180

**Comment by 01-vyom** [comment]
Sorry for the late reply, I think it's because the check is working on .rst files as well according to the include statement in the .yaml file:

https://github.com/pandas-dev/pandas/blob/2f49b8d32dc579358c63341730854f856ceb4315/.pre-commit-config.yaml#L96-L108

What should we do? separate this check with another id for only cython files?

**Comment by MarcoGorelli** [comment]
Ah yes, you're right. Then yes, just make another hook and only run it on `.pyx` and `.pxi.in` files in `pandas/_libs`

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -103,6 +103,13 @@ repos:
             |\.\.\ code-block\ ::
             |\.\.\ ipython\ ::
         types_or: [python, cython, rst]
+    -   id: cython-casting
+        name: Cython Casting
+        language: pygrep
+        entry: |
+            # Cython casting is of the form `<type>obj` as opposed to `<type> obj`
+            |[a-zA-Z0-9*]>\s
+        types: [cython]
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:112]: `pxi.in` isn't recognised as `cython` by [identify](https://github.com/pre-commit/identify), does the following work?

```suggestion
        files: (\.pyx|\.pxi.in)$
```

Comment (01-vyom) [reply .pre-commit-config.yaml:112]: I tried that, but it shows the following message, don't know if it worked or not as it showed the same message when I ran unwanted patterns with the cython check their. While, it gave error for .rst files when ran with ci-check on PR.

Screenshot 2021-08-02 at 8 51 19 PM

Comment (01-vyom) [reply .pre-commit-config.yaml:112]: Should I make a commit anyway?

Comment (MarcoGorelli) [reply .pre-commit-config.yaml:112]: ah, you should run `pre-commit run cython-casting --all-files`, else it'll only run against files you've staged

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -103,6 +103,13 @@ repos:
             |\.\.\ code-block\ ::
             |\.\.\ ipython\ ::
         types_or: [python, cython, rst]
+    -   id: cython-casting
+        name: Cython Casting
+        language: pygrep
+        entry: |
+            # Cython casting is of the form `<type>obj` as opposed to `<type> obj`
+            |[a-zA-Z0-9*]>\s
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:111]: this can probably just fit on one line: `entry: [a-zA-Z0-9*]>\s`

Comment (01-vyom) [reply .pre-commit-config.yaml:111]: I tried that, but I got the following error.

Screenshot 2021-08-02 at 10 00 55 PM

File: .pre-commit-config.yaml (hunk 2 above)
Comment (01-vyom) [reply .pre-commit-config.yaml:112]: Thank you!! I ran that and found that my hook is not working for the previous test.pyx or test.pix.in example, it works when I move the logic to unwanted hooks but it doesn't work with the new hook

File: .pre-commit-config.yaml (hunk 3 above)
Comment (MarcoGorelli) [reply .pre-commit-config.yaml:111]: ah you probably just need to quotes around the string

Comment (01-vyom) [reply .pre-commit-config.yaml:111]: Thank you it worked, will update the PR. Also, now I can just use the space at the end as before.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (01-vyom) [reply .pre-commit-config.yaml:112]: Ok with the quotes it works now

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -103,6 +103,11 @@ repos:
             |\.\.\ code-block\ ::
             |\.\.\ ipython\ ::
         types_or: [python, cython, rst]
+    -   id: cython-casting
+        name: Cython Casting
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:107]:
```suggestion
        name: Check Cython casting is `<type>obj`, not `<type> obj`
```

**Review by MarcoGorelli — APPROVED** [review]

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
sorry, forgot this last time, we should also remove 'lint' from this file (there's a few places where it crops up)

https://github.com/pandas-dev/pandas/blob/4f0e2e9712dd06d3e445c5777fd979c21cd84c02/ci/code_checks.sh#L13

**Comment by 01-vyom** [comment]
> sorry, forgot this last time, we should also remove 'lint' from this file (there's a few places where it crops up)
> 
> https://github.com/pandas-dev/pandas/blob/4f0e2e9712dd06d3e445c5777fd979c21cd84c02/ci/code_checks.sh#L13

Made the changes, should be good to go

**Comment by MarcoGorelli** [comment]
oops, CI checks are failing, there's still 

https://github.com/pandas-dev/pandas/blob/4f0e2e9712dd06d3e445c5777fd979c21cd84c02/.github/workflows/ci.yml#L51-L53

(after that, should be good to go!)

**Comment by 01-vyom** [comment]
Need to make changes in [docs](https://github.com/01-vyom/pandas/blob/pandas-42121/doc/source/development/contributing_codebase.rst#code-standards) as well.

**Comment by 01-vyom** [comment]
There are also docs for pattern matching, but it has been moved so should I change it as well?

[Doc1](https://github.com/01-vyom/pandas/blob/pandas-42121/doc/source/development/contributing_codebase.rst#code-standards)

Doc2: 

https://github.com/pandas-dev/pandas/blob/4f0e2e9712dd06d3e445c5777fd979c21cd84c02/ci/code_checks.sh#L5-L9

**Comment by MarcoGorelli** [comment]
Thanks for spotting that - yeah, sure!

**Review by MarcoGorelli — APPROVED** [review]

**Pull request merged**